cli/cmd/root.go

```diff
@@ -44,8 +44,11 @@ func addControlPlaneNetworkingArgs(cmd *cobra.Command) {
 	cmd.PersistentFlags().StringVar(&apiAddr, "api-addr", "", "Override kubeconfig and communicate directly with the control plane at host:port (mostly for testing)")
 }
 
-func newApiClient(kubeApi k8s.KubernetesApi) (pb.ApiClient, error) {
-	return public.NewExternalClient(controlPlaneNamespace, kubeApi)
+func newPublicAPIClient(api k8s.KubernetesApi, apiAddr string) (pb.ApiClient, error) {
```

Comment:
Cool, you're definitely on the right track here. I'd like to see this method written as:

```
func newPublicAPIClient() (pb.ApiClient, error) {
	if apiAddr != "" {
		return public.NewInternalClient(apiAddr)
	}

	kubeApi, err := k8s.NewK8sAPI(shell.NewUnixShell(), kubeconfigPath)
	if err != nil {
		return nil, err
	}

	return public.NewExternalClient(controlPlaneNamespace, kubeApi)
}
```

My rationale here is: we don't need to call `k8s.NewK8SAPI` if `apiAddr != ""`. So instead I'd prefer for this method to start with the CLI input strings, and only instantiate objects as needed. And if we're working with the CLI inputs (`apiAddr` and `kubeconfigPath`), then we don't need those passed in as parameters, since they're defined globally in this package.

The other change here is that we should need to pass `apiAddr` as a parameter to `k8s.NewK8sAPI`, since the apiAddr override has already been applied by the time we call that function. See my comments below for changes related to this.